### PR TITLE
Adding support for GenerateResourcesCodeAsConstants

### PR DESCRIPTION
**Pull request by vatsan-madhavan**
Adding support for (semi) self-managed resource-lookups by setting GenerateResourcesCodeAsConstants=true.

### CODE REVIEW

**Comment by vatsan-madhavan** [comment]
The hard coding of certain defaults, like `System`, `SR`, and notably `FxResources`, is a carryover from build-tools. I'd be happy to make these more configurable if need be.

File: src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs [hunk 1]
```diff
@@ -158,13 +195,34 @@ public override bool Execute()
...
                 {
                     case Lang.CSharp:
-                        strings.AppendLine($"{memberIndent}internal static string {identifier} => GetResourceString(\"{name}\"{defaultValue});");
+                        if (!AsConstants)
+                        {
+                            strings.AppendLine($"{memberIndent}internal static string {identifier} => GetResourceString(\"{name}\"{defaultValue});");
+                        }
+                        else
+                        {
+                            strings.AppendLine($"{memberIndent}internal static string {name} => nameof({name});");
```
Comment (rladuca) [inline src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs]: Is nameof buying you anything here over just using {name}?

Comment (vatsan-madhavan) [reply src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs]: In generated code, I doubt it. Is there a downside to it?

Comment (rladuca) [reply src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs]: I don't think so, the compiler maybe does a bit more work here to evaluate the nameof, but I doubt it's appreciable in any way.  I'm fine with either way really, just stuck out as strange.

File: src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs [hunk 2]
```diff
@@ -158,13 +195,34 @@ public override bool Execute()
                 switch (language)
                 {
                     case Lang.CSharp:
-                        strings.AppendLine($"{memberIndent}internal static string {identifier} => GetResourceString(\"{name}\"{defaultValue});");
+                        if (!AsConstants)
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs]: Nit: invert the condition.

File: src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs (hunk 1 above)
Comment (tmat) [reply src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs]: Why is this not a const field?

**Comment by ericstj** [comment]
Rather than add this back, did you consider using the properties and just using nameof(SR.Property) when you needed to get the resource key?  That's what CoreFx does and it results in much easier to read code.

The downside is it can create a bit of churn in the codebase, but its typically for the better.

File: src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs [hunk 3]
```diff
@@ -40,11 +46,24 @@ public sealed class GenerateResxSource : Task
         /// </summary>
         public string ResourceClassName { get; set; }
 
+        /// <summary>
+        /// Optionally, a namespace.type name in which the ResourceType property is generated. If not specified, defaults to System.SR. 
+        /// </summary>
+        /// <remarks>
+        /// This is ignored when <see cref="AsConstants"/> is false
+        /// </remarks>
+        public string SystemResourcesClassName { get; set; }
```
Comment (ericstj) [inline src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs]: ResourceName already defines the name of the resources and the locator type (EG: FxResources.) .  ResourceClassName already defines the name of the access class (EG: System.SR)

File: src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs [hunk 4]
```diff
@@ -280,6 +338,55 @@ public override bool Execute()
                 }
             }
 
+            string systemResourcesTypeDefinition = string.Empty;
```
Comment (ericstj) [inline src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs]: This is unnecessary.  You can use the existing support that generates this type and initializes the ResourceManager with it.  You'll need to change the partial class definition on the consumption side a little bit, but it shouldn't be a problem.  Here's what I did for CoreFX: https://github.com/dotnet/corefx/pull/32996/commits/169b36d13326eb9ba33389842ea97fcad774d22c

**Comment by ericstj** [comment]
> or now, I figured this was better than ~7k small edits

I see, back when we absorbed this change in CoreFx we did it as we moved code onto new infrastructure.  As such we were able to absorb those edits as part of the overall effort of moving the code.  If that doesn't apply then I see why you'd want to keep this functionality (that's the same reason I added the functionality to buildtools initially: to minimize edits).

I'm ok if you add the functionality here, but I believe you should be able to do so by minimally adding the `AsConstants`.

**Comment by vatsan-madhavan** [comment]
If there are no more concerns, can we merge this change? This would unblock ongoing work for our team. I'll be happy to make any additional changes if there is more feedback. cc @ericstj

**Comment by nguerrera** [comment]
Ping @markwilkie on whether this can be merged. Tomas is away, but his feedback was addressed.

File: src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs [hunk 5]
```diff
@@ -158,13 +167,34 @@ public override bool Execute()
...
+                        {
+                            strings.AppendLine($"{memberIndent}internal static string {identifier} => GetResourceString(\"{name}\"{defaultValue});");
+                        }
                         break;
 
                     case Lang.VisualBasic:
-                        strings.AppendLine($"{memberIndent}Friend Shared ReadOnly Property {identifier} As String");
+                        if (AsConstants)
+                        {
+                            strings.AppendLine($"{memberIndent}Friend Shared ReadOnly Property {name} As String");
```
Comment (ericstj) [inline src/Microsoft.DotNet.Arcade.Sdk/src/GenerateResxSource.cs]: Why does VB still implement these as properties when `AsConstants` is set?  I would have expected something like this: https://github.com/dotnet/buildtools/blob/61ba89d307f20d22c4e5a15a38f907b1246a6a9e/src/Microsoft.DotNet.Build.Tasks/GenerateResourcesCode.cs#L281

**Pull request merged**